File: app/src/main/java/com/sky/chowder/ui/adapter/LoaderURIAdapter.java
package com.sky.chowder.ui.adapter;

import android.animation.AnimatorInflater;
import android.graphics.Outline;
import android.os.Build;
import android.support.v7.widget.CardView;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.view.ViewOutlineProvider;
import android.widget.ImageView;

import com.sky.adapter.RecyclerAdapter;
import com.sky.adapter.RecyclerHolder;
import com.sky.chowder.R;
import com.sky.chowder.utils.ImageLoaderExecutors;

/**
 * @author sky QQ:1136096189
 * @Description:
 * @date 15/12/9 下午8:52
 */
public class LoaderURIAdapter extends RecyclerAdapter<String> {
    private ImageLoaderExecutors imageLoader;
    public String parentPath;

    public LoaderURIAdapter(int layoutId, String parentPath) {
        super(layoutId);
        this.parentPath = parentPath;
        imageLoader = new ImageLoaderExecutors();
    }

    public LoaderURIAdapter(int layoutId) {
        super(layoutId);
        imageLoader = new ImageLoaderExecutors();
    }

    public String getParentPath() {
        return parentPath;
    }

    public void setParentPath(String parentPath) {
        this.parentPath = parentPath;
    }

    @Override
    protected void onAchieveHolder(RecyclerHolder holder, int position) {

        CardView cardView = holder.getView(R.id.cardView);
        ImageView img = holder.getView(R.id.img_describe);
        img.setTag(parentPath + "/" + datas.get(position));
        img.setBackgroundResource(R.mipmap.ic_launcher);
        img.setMaxWidth(300);
        img.setMinimumWidth(300);
        imageLoader.loadImage(img, parentPath + "/" + datas.get(position));
        if (Build.VERSION.SDK_INT > Build.VERSION_CODES.KITKAT_WATCH) {
            //使用ripple
            cardView.setBackground(context.getDrawable(R.drawable.ripple));
            //点击效果，阴影效果
            cardView.setStateListAnimator(
                    AnimatorInflater.loadStateListAnimator(context, R.drawable.state_list_animator));
            //视图裁剪
            img.setClipToOutline(true);
            img.setOutlineProvider(new ViewOutlineProvider() {
                @Override
                public void getOutline(View view, Outline outline) {
                    outline.setRoundRect(view.getLeft(), view.getTop(),
                            view.getRight(), view.getBottom(), 30);
                }
            });
        }
    }

    /**
     * 空闲时在加载image
     *
     * @param start
     * @param last
     * @param viewGroup
     */
    public void setImageLoader(int start, int last, RecyclerView viewGroup) {
        for (int i = start; i <= last; i++) {
//            imageLoader.loadImage(viewGroup.findViewWithTag(parentPath + "/" + datas.get(i)), parentPath + "/" + datas.get(i));
        }
    }

    public void interruptExecutors() {
        imageLoader.closeExecutors();
    }

    public void getBitmap(int position) {

    }
}

File: app/src/main/java/com/sky/chowder/ui/adapter/LoaderUriAdapter.kt
package com.sky.chowder.ui.adapter

import android.animation.AnimatorInflater
import android.graphics.Outline
import android.os.Build
import android.support.annotation.RequiresApi
import android.support.v7.widget.RecyclerView
import android.view.View
import android.view.ViewOutlineProvider
import com.sky.adapter.RecyclerAdapter
import com.sky.adapter.RecyclerHolder
import com.sky.chowder.R
import com.sky.chowder.utils.imageloader.ImageLoaderExecutors
import kotlinx.android.synthetic.main.adapter_uri.view.*

/**
 * Created by SKY on 2015/12/9.
 */
class LoaderUriAdapter(layoutId: Int) : RecyclerAdapter<String>(layoutId) {
    private val imageLoader: ImageLoaderExecutors = ImageLoaderExecutors()
    var parentPath: String? = null

    override fun onAchieveHolder(holder: RecyclerHolder, position: Int) {
        with(holder!!.itemView) {
            image.tag = parentPath + "/" + datas[position]
            image.setBackgroundResource(R.mipmap.ic_launcher)
            image.maxWidth = 300
            image.minimumWidth = 300
//            imageLoader.loadImage(image, "$parentPath/${datas[position]}")
            if (Build.VERSION.SDK_INT > Build.VERSION_CODES.KITKAT_WATCH) {
                //使用ripple
                cardView.background = context.getDrawable(R.drawable.ripple)
                //点击效果，阴影效果
                cardView.stateListAnimator = AnimatorInflater.loadStateListAnimator(context, R.drawable.state_list_animator)
                //视图裁剪
                image.clipToOutline = true
                image.outlineProvider = object : ViewOutlineProvider() {
                    @RequiresApi(Build.VERSION_CODES.LOLLIPOP)
                    override fun getOutline(view: View, outline: Outline) {
                        outline.setRoundRect(view.left, view.top, view.right, view.bottom, 30f)
                    }
                }
            }
        }
    }

    /**
     * 空闲时在加载image
     *
     * @param start
     * @param last
     * @param viewGroup
     */
    fun setImageLoader(start: Int, last: Int, viewGroup: RecyclerView) {
        for (i in start..last) {
            imageLoader.loadImage(viewGroup.findViewWithTag("$parentPath/${datas[i]}"), "$parentPath/${datas[i]}");
        }
    }

    fun interruptExecutors() {
        imageLoader.closeExecutors()
    }
}